js/server/modules/@arangodb/foxx/manager.js

```diff
@@ -856,94 +926,140 @@ function uninstall (mount, options = {}) {
   return oldService;
 }
 
-function replace (serviceInfo, mount, options = {}) {
-  ensureFoxxInitialized();
-  if (!options.force) {
-    const serviceDefinition = utils.getServiceDefinition(mount);
-    if (!serviceDefinition) {
-      throw new ArangoError({
-        errorNum: errors.ERROR_SERVICE_NOT_FOUND.code,
-        errorMessage: dd`
+function replace(serviceInfo, mount, options = {}) {
+  const locked = _checkServiceModificationLock();
+  if (locked) {
+    // TODO: Throw if we were not able to get the lock within 10 seconds
+    return;
+  }
+
+  try {
+    _setServiceModificationLock(true);
+    ensureFoxxInitialized();
+    if (!options.force) {
+      const serviceDefinition = utils.getServiceDefinition(mount);
+      if (!serviceDefinition) {
+        throw new ArangoError({
+          errorNum: errors.ERROR_SERVICE_NOT_FOUND.code,
+          errorMessage: dd`
           ${errors.ERROR_SERVICE_NOT_FOUND.message}
           Mount path: "${mount}".
         `
+        });
+      }
+    } else {
+      utils.validateMount(mount);
+    }
+    const tempPaths = _prepareService(serviceInfo, options.legacy);
+    const tempService = FoxxService.create({
+      mount,
+      basePath: tempPaths.tempServicePath,
+      noisy: true,
+      options: {
+        development: options.development,
+        configuration: options.configuration,
+        dependencies: options.dependencies
+      }
+    });
+
+    // TODO: Maybe find a more elegant solution here
+    // Set lock to false, because uninstall & install are
+    // taking care of their locks
+
+    try {
+      _setServiceModificationLock(false);
+      _uninstall(mount, {
+        teardown: options.teardown !== false,
+        force: true
+      });
+      _install(tempService, tempPaths.tempBundlePath, {
+        setup: options.setup !== false,
+        force: true
       });
+    } finally {
+      _setServiceModificationLock(false);
```

Comment:
If the lock has been set to false that means an install could have started on a different thread in the meantime. This would release the lock a parallel install might have acquired, causing havoc.